docs/internal_abi.md

```diff
@@ -0,0 +1,323 @@
+The alignment of the result is the alignment of the most strongly aligned type.
+
+If one of the types is a pointer, the object size is pointer sized, and the discriminator is three or fewer bits, then the low bits of the pointer are used for the discriminator.
+Pointer values are stored with their bits in the natural positions and must have the low bits subtracted away on use.
+Other values are stored in the most significant bits.
+
+*Rationale:* Most ISAs make subtracting a 3-bit immediate trivial (often something that can be folded into another instruction).
+Other values that fit in the payload space can be shifted down as required in a single-cycle instruction.
+
+If the payload is an integer number of bytes, the payload is stored first (independent of byte order) so that the payload can benefit from strong alignment, which avoids the payload accidentally spanning a cache line.
+In this case, all elements of the payload are stored starting from the beginning of the payload.
+
+*Rationale:* The payload can be loaded with a single aligned load on any architecture and depending on the target byte order then needs to do either a single zero-extending shift right or a left shift followed by a zero-extending right shift.
+
+If storing the discriminator bits at either end of the space results in different numbers of complete bytes for the payload then the discriminator should stored at the end that results in the larger number of complete bytes for the payload.
+If there are no discriminator bits, this rule is applied as if there were a single bit.
+Payload values that are smaller than the payload space are then stored aligned to the opposite end of the value to the discriminator.
+
+*Rationale:* This case is reached if we are packing a union type inside the significand bits of a signalling NaN and guarantees that we can load the payload in a single aligned load and need to zero bits at only one end to extract the payload.
+
+In other cases, the discriminator bits are stored in the low bits (as defined by the platform byte order) of the object and the payload in the high bits.
+
+*Rationale:* The discriminator is easy to mask off with a right shift and extract with an xor with an immediate operand for comparisons.
+
+The types identified in a discriminator are sorted lexically and discriminator values are assigned in order.
+Singletons are sorted before all other types.
+
+*Rationale:* The representation of `A | B` and `B | A` should be the same.
+It is useful to sort singletons before pointers so that the common case of `T* | NotAThing` is able to use null-pointer checks and branches, which are typically optimised aggressively in modern microarchitectures.
+
+{*Note:* I'm not sure lexical sorting will work because types imported from the same package by two other packages will have different names and so we may need some other stable ordering for types.}
```

Comment:
I believe all types will be normalised by their full scope, so even if you name them differently from different modules, they should, at the end, have the same canonical qualified name.